packages/base64/src/decode.js

```diff
@@ -18,46 +18,55 @@ import { monodu64, padding } from './common.js';
  * messages.
  * @returns {Uint8Array} decoded bytes
  */
-const jsDecodeBase64 = (string, name = '<unknown>') => {
-  const data = new Uint8Array(Math.ceil((string.length * 4) / 3));
+export const jsDecodeBase64 = (string, name = '<unknown>') => {
+  // eslint-disable-next-line no-nested-ternary
+  const paddingLength = string.endsWith('==')
+    ? 2
+    : string.endsWith('=')
+    ? 1
+    : 0;
+  const data = new Uint8Array((string.length * 3) / 4 - paddingLength);
```

Comment:
I'm confused, why check the end of the string? `new Uint8Array(Math.ceil((string.length * 3) / 4))` should be correct.

Checking the end of the string has 2 issues: it disallows trimming the trailing `=` which is a very common case, and I suspect `endWith` may be `O(n)` as well.

On that note. `string.length` is `O(n)` and looked up twice in this implementation.

Reply:
> I'm confused, why check the end of the string? `new Uint8Array(Math.ceil((string.length * 3) / 4))` should be correct.

"Cg==", "QQo=", and "QUIK" all have length 4, but encode different-length data (respectively, 1, 2, and 3 octets). Trailing padding is necessary to differentiate them.

> Checking the end of the string has 2 issues: it disallows trimming the trailing `=` which is a very common case,

...but not one supported by this code.

> and I suspect `endWith` may be `O(n)` as well.
> 
> On that note. `string.length` is `O(n)` and looked up twice in this implementation.

Hrm, I thought not but I just double-checked and you're right about both. Updated.